node/bindnode/node.go

```diff
@@ -0,0 +1,1099 @@
+}
+
+func (w *_prototype) NewBuilder() ipld.NodeBuilder {
+	return &_builder{_assembler{
+		schemaType: w.schemaType,
+		val:        reflect.New(w.goType).Elem(),
+	}}
+}
+
+// TODO: consider these Typed interfaces for the schema package
+
+type TypedPrototype interface {
+	ipld.NodePrototype
+
+	Representation() ipld.NodePrototype
+}
+
+type TypedAssembler interface {
+	ipld.NodeAssembler
+
+	Representation() ipld.NodeAssembler
+}
+
+func (w *_prototype) Representation() ipld.NodePrototype {
+	return (*_prototypeRepr)(w)
+}
+
+var (
+	goTypeString = reflect.TypeOf("")
+	goTypeBytes  = reflect.TypeOf([]byte{})
+	goTypeLink   = reflect.TypeOf((*ipld.Link)(nil)).Elem()
+
+	schemaTypeFieldName = schema.SpawnString("fieldNameString")
+)
+
+type _node struct {
+	schemaType schema.Type
+
+	val reflect.Value // non-pointer
+}
+
+// TODO: only expose TypedNode methods if the schema was explicit.
+// type _typedNode struct {
+// 	_node
+// }
+
+func (w *_node) Type() schema.Type {
+	return w.schemaType
+}
+
+func (w *_node) Representation() ipld.Node {
+	return (*_nodeRepr)(w)
+}
+
+func (w *_node) Kind() ipld.Kind {
+	return w.schemaType.TypeKind().ActsLike()
+}
+
+func (w *_node) LookupByString(key string) (ipld.Node, error) {
+	switch typ := w.schemaType.(type) {
+	case *schema.TypeStruct:
+		field := typ.Field(key)
+		if field == nil {
+			return nil, ipld.ErrInvalidKey{
+				TypeName: typ.Name().String(),
+				Key:      basicnode.NewString(key),
+			}
+		}
+		fval := w.val.FieldByName(fieldNameFromSchema(key))
+		if !fval.IsValid() {
+			panic("TODO: go-schema mismatch")
+		}
+		if field.IsOptional() {
+			if fval.IsNil() {
+				return ipld.Absent, nil
+			}
+			fval = fval.Elem()
+		}
+		if field.IsNullable() {
+			if fval.IsNil() {
+				return ipld.Null, nil
+			}
+			fval = fval.Elem()
+		}
+		node := &_node{
+			schemaType: field.Type(),
+			val:        fval,
+		}
+		return node, nil
+	case *schema.TypeMap:
+		var kval reflect.Value
+		switch ktyp := typ.KeyType().(type) {
+		case *schema.TypeString:
+			kval = reflect.ValueOf(key)
+		default:
+			asm := &_assembler{
+				schemaType: ktyp,
+				val:        reflect.New(w.val.Type().Key()).Elem(),
+			}
+			if err := (*_assemblerRepr)(asm).AssignString(key); err != nil {
+				return nil, err
+			}
+			kval = asm.val
+		}
+		fval := w.val.MapIndex(kval)
+		if !fval.IsValid() { // not found
+			return nil, ipld.ErrNotExists{
+				// TODO
+			}
+		}
+		if typ.ValueIsNullable() {
+			if fval.IsNil() {
+				return ipld.Null, nil
+			}
+			fval = fval.Elem()
+		}
+		node := &_node{
+			schemaType: typ.ValueType(),
+			val:        fval,
+		}
+		return node, nil
+	case *schema.TypeUnion:
+		var idx int
+		var mtyp schema.Type
+		for i, member := range typ.Members() {
+			if member.Name().String() == key {
+				idx = i
+				mtyp = member
+				break
+			}
+		}
+		if mtyp == nil { // not found
+			return nil, ipld.ErrNotExists{
+				// TODO
+			}
+		}
+		haveIdx := int(w.val.FieldByName("Index").Int())
```

Comment:
This would work if we assume we always infer the type, but longer term I want to allow the user more control. You can imagine allowing the user to provide a custom type to back a union by providing methods like `Get() (tag int, value interface{})` and `Put(tag int, value interface{})`.